packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,384 @@
+          removeFromParent(value[i]);
+        }
+      } else if (value != null) {
+        // TemplateInstance case: If the value has disconnectable children (will
+        // only be in the case that it is a TemplateInstance), we disconnect it
+        // and remove it from this NodePart's disconnectable children
+        setChildrenConnected(value as DisconnectableParent, isConnected);
+        removeFromParent(value as DisconnectableParent);
+      }
+    } else {
+      setChildrenConnected(this, isConnected);
+    }
+  }
+}
+
+/**
+ * Patches disconnection API onto NodeParts.
+ */
+const installDisconnectAPI = (child: DisconnectableParent) => {
+  if ((child as NodePart).type == NODE_PART) {
+    (child as NodePart)._$setNodePartConnected ??= setNodePartConnected;
+  }
+};
+
+/**
+ * An abstract `Directive` base class whose `disconnectedCallback` will be
+ * called when the part containing the directive is cleared as a result of
+ * re-rendering, or when the user calls `part.setDirectiveConnection(false)` on
+ * a part that was previously rendered containing the directive.
+ *
+ * If `part.setDirectiveConnection(true)` is subsequently called on a containing
+ * part, the directive's `reconnectedCallback` will be called prior to its next
+ * `update`/`render` callbacks. When implementing `disconnectedCallback`,
+ * `reconnectedCallback` should also be implemented to be compatible with
+ * reconnection.
+ */
+export abstract class DisconnectableDirective extends Directive {
+  isConnected = true;
+  _pendingValue: unknown = noChange;
+  _$disconnetableChildren?: Set<DisconnectableParent> = undefined;
+  constructor(partInfo: PartInfo) {
+    super(partInfo);
+    this._$parent = partInfo._$parent;
+    // Climb the parent tree, creating a sparse tree of children needing
+    // disconnection
+    for (
+      let current = this as DisconnectableParent, parent;
+      (parent = current._$parent);
+      current = parent
+    ) {
+      let children = parent._$disconnetableChildren;
+      if (children === undefined) {
+        parent._$disconnetableChildren = children = new Set();
+      } else if (children.has(current)) {
+        // Once we've reached a parent that already contains this child, we
+        // can short-circuit
+        break;
+      }
+      children.add(current);
+      installDisconnectAPI(parent);
+    }
+  }
+  /**
+   * Called from the core code when a directive is going away from a part (in
+   * which case `shouldRemoveFromParent` should be true), and from the
+   * `setChildrenConnected` helper function when recursively changing the
+   * connection state of a tree (in which case `shouldRemoveFromParent` should
+   * be false).
+   *
+   * @param isConnected
+   * @param shouldRemoveFromParent
+   * @internal
+   */
+  _$setDirectiveConnected(isConnected: boolean, shouldRemoveFromParent = true) {
+    this._setConnected(isConnected);
+    setChildrenConnected(this, isConnected);
```

Comment:
If you move `setChildrenConnected(this, isConnected);` into the condition below, you can simplify `setChildrenConnected`

Reply:
Done